src/commands/BootstrapCommand.js

```diff
@@ -14,147 +14,345 @@ export default class BootstrapCommand extends Command {
+      // and is not already installed at root
+      if (!hasPackage(name) && !this.dependencySatisfiesPackages(name, deps.map(dep => findPackage(dep)))) {
+        // add the common version to root install
+        installs.__ROOT__.push(`${name}@${commonVersion}`);
+      }
+      // add less common versions to package installs
+      allVersions.forEach(version => {
+        // only install less common versions,
+        // unless it's a version-mismatched package
+        if (version !== commonVersion || hasPackage(name)) {
+          depsToInstall[name].dependents[version].forEach(pkg => {
+            // only install dependency if it's not already installed
+            if (!findPackage(pkg).hasDependencyInstalled(name)) {
+              if (!installs[pkg]) {
+                installs[pkg] = [];
+              }
+              installs[pkg].push(`${name}@${version}`);
+              this.logger.warning(
+                `"${pkg}" package depends on ${name}@${version}, ` +
+                `which differs from the more common ${name}@${commonVersion}.`
+              );
+            }
+          });
+        }
       });
-
-    if (externalPackages.length) {
-      NpmUtilities.installInDir(pkg.location, externalPackages, callback);
-    } else {
-      callback();
-    }
+    });
+    return installs;
   }
 
-  hasMatchingDependency(pkg, dependency, showWarning = false) {
-    const expectedVersion = pkg.allDependencies[dependency.name];
-    const actualVersion = dependency.version;
-
-    if (!expectedVersion) {
+  /**
+   * Determine if a dependency installed at the root satifies the requirements of the passed packages
+   * This helps to optimize the bootstrap process and skip dependencies that are already installed
+   * @param {String} dependency
+   * @param {Array.<String>} packages
+   */
+  dependencySatisfiesPackages(dependency, packages) {
+    const packageJson = path.join(this.repository.rootPath, "node_modules", dependency, "package.json");
+    try {
+      return packages.every(pkg => {
+        return semver.satisfies(
+          require(packageJson).version,
+          pkg.allDependencies[dependency]
+        );
+      });
+    } catch (e) {
       return false;
     }
+  }
 
-    if (this.isCompatableVersion(actualVersion, expectedVersion)) {
-      return true;
-    }
-
-    if (showWarning) {
-      this.logger.warning(
-        `Version mismatch inside "${pkg.name}". ` +
-        `Depends on "${dependency.name}@${expectedVersion}" ` +
-        `instead of "${dependency.name}@${actualVersion}".`
-      );
+  /**
+   * Install external dependencies for packages
+   * @param {Object} dependencies
+   * @param {Function} callback
+   */
+  installExternalDependencies(dependencies, callback) {
+    const actions = [];
+    Object.keys(dependencies)
+      .forEach(dest => {
+        const destLocation = dest === "__ROOT__" ?
+          this.repository.rootPath :
```

Comment:
This might be safer to put into `this.repository.packagesLocation`.  It's possible that the repo root has a dependency on a package from within the repo, which would conflict with the symlink.  If the `__ROOT__` deps went into `packages/` then they would have a clean directory with no potential for conflict.